### PR TITLE
add auth and registration

### PR DESCRIPTION
**Pull request by v1tal4ik**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+.env.local
+.env.development.local
+.env.test.local
+.env.production.local
+
+package-lock.json*
+npm-debug.log*
+yarn-debug.log*
+yarn-error.log*
+yarn.lock*
```
Comment (vickierose) [inline .gitignore]: I'd better leave yarn.lock. It works fine so there's no need to keep it out of working tree

File: src/components/Header/Header.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    if(this.props.user.isAuth){
+      this.setState({ userAuth: true });
+    }
+  }
+
+  componentDidMount = () =>{
+    store.subscribe(this.userAuthObserver);
+  }
+
+  foo = () =>{
```
Comment (vickierose) [inline src/components/Header/Header.js]: wtf is this?

File: src/components/Header/Header.js [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    console.log('work');
+  }
+
+
+  render() {
+    const { userAuth } = this.state;
+    return (
+      <header>
```
Comment (vickierose) [inline src/components/Header/Header.js]: `<header>` is not a thing you'd want to use in `body`. Replace it with h1 or some other heading tag. info about `header`: https://www.w3schools.com/tags/tag_head.asp
https://developer.mozilla.org/ru/docs/Learn/HTML/%D0%92%D0%B2%D0%B5%D0%B4%D0%B5%D0%BD%D0%B8%D0%B5_%D0%B2_HTML/The_head_metadata_in_HTML

File: src/components/Header/Header.js [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+  }
+}
+
+
+export default connect(state => ({
+  user: getUser(state),
+}), {})(Header);
+
+
+// <Link to='/auth/login'><span>Logout</span></Link>
```
Comment (vickierose) [inline src/components/Header/Header.js]: putting commented code without explanations into main branch is bad practice. Remove it or add a comment why you need to paste it in here

File: src/components/Login/Login.js [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+          e.preventDefault();
+          const result = await singInById({ id });
+          const path = result ? '/': 'login';
+          user.isAuth = true;
+          await userAuth(user);
+          this.props.history.push(path);
+        }
+      };
+      
+
```
Comment (vickierose) [inline src/components/Login/Login.js]: too many empty lines. Watch for this! 1 line is enough

File: src/components/Login/Login.js [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+                <input className="login-form-input" type="text" name='email' placeholder='Example@gmail.com' onChange={this.handleChangeInput} onBlur={this.checkEmail} required/>
+                {check.email ?
+                        <div className = 'form-status-icon' > 
+                            {status.email ?  <i className="fas fa-check green"></i > : < i className = "fas fa-times red" > </i>} 
+                        </div> :
+                    null
+                }
+
+                <label className='login-form-label'>Password</label>
+                <input className="login-form-input" type="password" name='password' onChange={this.handleChangeInput} onBlur={this.checkPassword.bind(this)} />
```
Comment (vickierose) [inline src/components/Login/Login.js]: `checkPassword` is an arrow function, you don't need to bind it. And even if you need to bind anything you'd better do it in constructor

File: src/components/Market/Market.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+import './style.css';
+
+
+class Market extends Component {
+  constructor(props) {
+    super();
+
+  }
+
+  test = async () => {
```
Comment (vickierose) [inline src/components/Market/Market.js]: wtf is this?

File: src/components/Register/Register.js [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+      msg: '',
+    };
+  }
+
+    handleChangeInput = ({ target: { name, value } }) => {
+      this.setState({ data: { ...this.state.data, [name]: value } });
+    };
+
+    checkEmail = async ({ target: { value: email } }) => {
+      const exp = /@/;
```
Comment (vickierose) [inline src/components/Register/Register.js]: I'd better use more specific regexp. You can find a ton of them

File: src/components/Register/Register.js [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+    const {
+      data, check, status, visible, msg,
+    } = this.state;
+    return (
+            <>
+            <form className='register-form'>
+                <p className='login-form-title'>Register</p>
+
+                <label className='login-form-label'>Email <span>(unique)</span></label>
+                <input className="login-form-input" type="text" name='email' value={data.email} placeholder='Example@gmail.com' onChange={this.handleChangeInput} onBlur={this.checkEmail.bind(this)} required/>
```
Comment (vickierose) [inline src/components/Register/Register.js]: already made a note about bind in prev files

Comment (vickierose) [inline src/components/Register/Register.js]: also for better readability I'd suggest to split this line into multiple like so:
```
<input 
  className="login-form-input"
  type="text"
  name='email'
  value={data.email}
  placeholder='Example@gmail.com'
  onChange={this.handleChangeInput}
  onBlur={this.checkEmail.bind(this)}
  required
/>
```
common practice is to use this approach if you're passing more than 3 props or attributes to component

File: src/components/Register/Register.js [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+                {check.email
+                  ? <div className = 'form-status-icon' >
+                            {status.email ? <i className="fas fa-check green"></i > : < i className = "fas fa-times red" > </i>}
```
Comment (vickierose) [inline src/components/Register/Register.js]: lines of this type also can be splitted:
```
{
  status.email ? 
  <i className="fas fa-check green"></i > : 
  < i className = "fas fa-times red" > </i>
}
```

File: src/modules/index.js [hunk 11]
```diff
@@ -0,0 +1,11 @@
+import { combineReducers } from 'redux';
+// import { fork } from 'redux-saga/effects';
```
Comment (vickierose) [inline src/modules/index.js]: already made a note about commented code

File: src/modules/user/actions.js [hunk 12]
```diff
@@ -0,0 +1,5 @@
+import { createAction } from 'redux-actions';
+
```
Comment (vickierose) [inline src/modules/user/actions.js]: I'd prefer you to use vanilla redux so you could understand how it works on practice

File: src/modules/user/reducer.js [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+  [fetchUserDataSuccess]: (_state, action) => action.payload,
+  [fetchUserDataFailure]: (_state, action) => action.payload,
+  [userAuth]: (_state, action) => action.payload,
+}, {});
+
+
+export default user;
+
+
+export const getUser = state => state.user;
```
Comment (vickierose) [inline src/modules/user/reducer.js]: these functions are called selectors and should go to a separate file called `selectors.js`

File: src/modules/user/reducer.js [hunk 14]
```diff
@@ -0,0 +1,19 @@
+import { handleActions } from 'redux-actions';
+import {
+  fetchUserDataSuccess,
+  fetchUserDataFailure,
+  userAuth,
+} from './actions';
+
+
+const user = handleActions({
```
Comment (vickierose) [inline src/modules/user/reducer.js]: also would prefer you to use vanilla redux

File: src/components/Register/Register.js [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+  render() {
+    const {
+      data, check, status, visible, msg,
+    } = this.state;
+    return (
+            <>
+            <form className='register-form'>
+                <p className='login-form-title'>Register</p>
+
+                <label className='login-form-label'>Email <span>(unique)</span></label>
```
Comment (vickierose) [inline src/components/Register/Register.js]: these label-input-status groups could go to a separate common component as they have common styles and functionality.

File: src/components/Header/Header.js (hunk 3 above)
Comment (v1tal4ik) [reply src/components/Header/Header.js]: I use tag <header>, not <head>.
http://htmlbook.ru/html/header

Comment (v1tal4ik) [reply src/components/Header/Header.js]: I use tag <header>, not <head>. Maybe you mixed up? http://htmlbook.ru/html/header

Comment (vickierose) [reply src/components/Header/Header.js]: oh, sorry, you're right 😅

File: src/components/Login/Login.js (hunk 6 above)
Comment (vickierose) [reply src/components/Login/Login.js]: still relevant

**Pull request merged**